wellcomeml/ml/cnn.py

```diff
@@ -98,9 +112,53 @@ def _get_distributed_strategy(self):
             strategy = tf.distribute.get_strategy()
         return strategy
 
+    def _init_from_data(self, X, Y):
+        logger.info(
+            "Initializing sequence_length, vocab_size \
+            and nb_outputs from data. This might take a while."
+        )
+        if type(X) is np.ndarray:
+            X_max = X.max()
+            Y_max = Y.max()
+            X_shape = X.shape[1]
+            Y_shape = Y.shape[1] if self.multilabel else Y.shape[0]
+            steps_per_epoch = math.ceil(X.shape[0] / self.batch_size)
+        elif isinstance(X, tf.data.Dataset):
+            X = X.batch(self.batch_size)
+
+            # init from iterating over dataset
+            X_max = 0
+            Y_max = 0
+            X_shape = None
+            Y_shape = None
+            steps_per_epoch = 0
+            for X_batch, Y_batch in X:
+                batch_X_max = X_batch.numpy().max()
```

Comment:
I don't understand this `.numpy().max()` operation..

Reply:
Does this give you the length? I thought that's what you needed to calculate

Reply:
it gives me the maximum integer in X batch. I want to find the vocabulary size.

the length of the columns is also used to define sequence length but the variable is called X_shape.

happy to rename if anything is unclear